gensim/similarities/docsim.py

```diff
@@ -186,11 +186,27 @@ def __init__(self, output_prefix, corpus, num_features, num_best=None, chunksize
         `self.num_best = 10` before doing a query.
 
         """
+        self.is_shard = False  # To indicate if shard directory has been created, if not then shard and index files are stored in temp
         if output_prefix is None:
             # undocumented feature: set output_prefix=None to create the server in temp
             self.output_prefix = utils.randfname(prefix='simserver')
         else:
-            self.output_prefix = output_prefix
+            try:
+                self.output_prefix = os.path.join(os.path.dirname(output_prefix), 'shard', '')   # '' to get the trailing slash
+                if not os.path.exists(self.output_prefix):
+                    os.makedirs(self.output_prefix)
+                    logger.info("created %s directory . Both shard and index files are stored at this location. Please use \
+                    this entire folder for portability across different systems. ", self.output_prefix)
+                    self.is_shard = True  # helps while saving the index file as we don't want to create directories in temp
+                else:
+                    logger.warning("It seems there is already a file named shard in location %s . Please ensure  \
+                    that there is no other files named shard in the provided location. Starting similarity index under temp folder", os.path.join(os.path.dirname(output_prefix), ''))
+                    self.output_prefix = utils.randfname(prefix='simserver')
+            except:
+                logger.warning("Failed to create shard folder in %s. Please ensure that you have required permission to create  \
```

Comment:
yes, this API concern is causing delay for this PR. As mentioned in the description, as of now, we ask user for `output_prefix` location where we save shards as `out_prefix.shard_number`, and while saving we ask for location for `index object pickle`. This causes problem while loading, if user has moved pickle to a different system where they don't have shards, or if user has changed the location of `shards`.

I understand this approach of creating a directory is quite ambiguous, so do you think it will be better to ask the user for a location like `/home/prakhar/index` and create a directory `index` in `/home/peakhar/` and inside that directory create shards as `index.0` etc. And while saving too, we only ask for pickle name and not entire location, which we will save inside that directory only. The catch here, however, is what if user provides something like `/home/prakhar/` only. As of now, we would create shards `.0`, `.1` etc, but how would we create a directory then ?

After all, we have to save as `pickle_path = output_prefix + index.pkl`